core/server.go

```diff
@@ -657,25 +657,43 @@ func (m *RpcServer) SetRepoDefaultRevision(cxt context.Context, req *pb.SetRepoD
 
 func (m *RpcServer) Subscribe(req *pb.SubscribeRequest, srv pb.Control_SubscribeServer) error {
 	m.logMethod()
-	ctx := srv.Context()
+	ch := make(chan *pb.Event)
+	// use to subscribe to taskman
+	// here just do it
+	sub := m.state.PublicEventFeed.Subscribe(ch)
 	for {
 		select {
-		case event := <- m.state.PublicEvent:
-			err := srv.Send(event)
-			if err != nil {
-				return err
+		case event, ok := <- ch:
+			if ok {
+				err := srv.Send(event)
+				if err != nil {
+					return err
+				}
 			}
-		case <-ctx.Done():
-            return ctx.Err()
+		case err := <- sub.Err():
+            return err
 		}
 	}
 }
 
-func (m *RpcServer) NewAutoEnvironment(cxt context.Context, request *pb.NewEnvironmentRequest) (*pb.NewAutoEnvironmentReply, error) {
+func (m *RpcServer) NewAutoEnvironment(request *pb.NewEnvironmentRequest, srv pb.Control_NewAutoEnvironmentServer) error {
 	m.logMethod()
-	go m.state.environments.CreateAutoEnvironment(request.GetWorkflowTemplate(), request.GetVars())
-	r := &pb.NewAutoEnvironmentReply{
+	ch := make(chan *pb.Event)
+	// use to subscribe to taskman
+	// here just do it
+	sub := m.state.PublicEventFeed.Subscribe(ch)
+	go m.state.environments.CreateAutoEnvironment(request.GetWorkflowTemplate(), request.GetVars(), sub)
+	for {
+		select {
+		case event, ok := <- ch:
+			if ok {
+				err := srv.Send(event)
```

Comment:
We still have the option of sending out Role events instead of task events, but in this case we could consider emitting Task events too with relative ease. Assuming the `publicEventFeed` gives us *all* task events of the system, we can simply check whether the event's task's environment id equals that of the current environment, and if yes, forward it. It would require some changes to `CreateAutoEnvironment` I assume, because you'd need to find out the new environment id, but it could be a reasonable way. Let's discuss.